File: components/digid_dws/src/main/java/nl/logius/digid/dws/client/RdwClient.java

/*
  Deze broncode is openbaar gemaakt vanwege een Woo-verzoek zodat deze
  gericht is op transparantie en niet op hergebruik. Hergebruik van 
  de broncode is toegestaan onder de EUPL licentie, met uitzondering 
  van broncode waarvoor een andere licentie is aangegeven.
  
  Het archief waar dit bestand deel van uitmaakt is te vinden op:
    https://github.com/MinBZK/woo-besluit-broncode-digid
  
  Eventuele kwetsbaarheden kunnen worden gemeld bij het NCSC via:
    https://www.ncsc.nl/contact/kwetsbaarheid-melden
  onder vermelding van "Logius, openbaar gemaakte broncode DigiD" 
  
  Voor overige vragen over dit Woo-besluit kunt u mailen met open@logius.nl
  
  This code has been disclosed in response to a request under the Dutch
  Open Government Act ("Wet open Overheid"). This implies that publication 
  is primarily driven by the need for transparence, not re-use.
  Re-use is permitted under the EUPL-license, with the exception 
  of source files that contain a different license.
  
  The archive that this file originates from can be found at:
    https://github.com/MinBZK/woo-besluit-broncode-digid
  
  Security vulnerabilities may be responsibly disclosed via the Dutch NCSC:
    https://www.ncsc.nl/contact/kwetsbaarheid-melden
  using the reference "Logius, publicly disclosed source code DigiD" 
  
  Other questions regarding this Open Goverment Act decision may be
  directed via email to open@logius.nl
*/

package nl.logius.digid.dws.client;

import nl.logius.digid.dws.exception.PenRequestException;
import nl.logius.digid.dws.exception.PukRequestException;
import nl.rdw.eid_wus_crb._1.*;
import org.apache.cxf.endpoint.Client;
import org.apache.cxf.frontend.ClientProxy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.xml.ws.BindingProvider;
import java.util.HashMap;
import java.util.Map;

@Component
public class RdwClient extends AbstractWsClient {

    @Autowired
    public RdwClient(@Value("${ws.client.connection_timeout}") Integer connectionTimeout,
                       @Value("${ws.client.receive_timeout}") Integer receiveTimeout,
                       @Value("${rdw_private_key_password}") String clientTlsPassphrase,
                       @Value("${rdw_client_tls_keystore}") String clientTlsKeystore) {
        super(connectionTimeout, receiveTimeout, clientTlsPassphrase, clientTlsKeystore );
    }

    @Value("${ws.client.rdw_endpoint_url}")
    private String endpoint;

    private final Logger logger = LoggerFactory.getLogger(getClass());

    public PenAanvraagEIDResponse penRequest(PenAanvraagEIDRequest rdwRequest) throws PenRequestException {
        final IEidOpRijbewijs port = setupClient();

         try {
            final PenAanvraagEIDResponse penAanvraagResponse = port.getPEN(rdwRequest);
            if (penAanvraagResponse.getPenAanvraagEIDFault() != null) {
                throw new PenRequestException(penAanvraagResponse.getPenAanvraagEIDFault());
            }
            return penAanvraagResponse;
        } catch (javax.xml.ws.WebServiceException e) {
            logger.error("During the PEN request call to RDW a javax.xml.ws.WebServiceException has been thrown");
            throw new PenRequestException("DWS4", e);
        }
    }

    public OpvragenPUKCodeEIDResponse pukRequest(OpvragenPUKCodeEIDRequest rdwRequest) throws PukRequestException {
        final IEidOpRijbewijs port = setupClient();

        try {
            final OpvragenPUKCodeEIDResponse opvragenPUKCodeEIDResponse = port.getPUK(rdwRequest);
            if (opvragenPUKCodeEIDResponse.getOpvragenPUKCodeEIDFault() != null) {
                throw new PukRequestException(opvragenPUKCodeEIDResponse.getOpvragenPUKCodeEIDFault());
            }
            return opvragenPUKCodeEIDResponse;
        } catch (javax.xml.ws.WebServiceException e) {
            logger.error("During the PEN request call to RDW a javax.xml.ws.WebServiceException has been thrown");
            throw new PukRequestException("DWS4", e);
        }
    }


    private IEidOpRijbewijs setupClient() {
        final EidOpRijbewijs service = new EidOpRijbewijs();
        final IEidOpRijbewijs port = service.getCustomBindingIEidOpRijbewijs();

        final BindingProvider bp = (BindingProvider)port;
        bp.getRequestContext().put(BindingProvider.ENDPOINT_ADDRESS_PROPERTY, endpoint);

        final Map<String, String> nsMap = new HashMap<>();
        nsMap.put("wsa", "http://www.w3.org/2005/08/addressing"); // add wsa namespace to envelope
        final Client client = ClientProxy.getClient(port);
        client.getRequestContext().put("soap.env.ns.map", nsMap);

        setTimeouts(client);
        setupTLS(client);
        return port;
    }


}

File: components/digid_x/app/controllers/concerns/rdw_client.rb

# Deze broncode is openbaar gemaakt vanwege een Woo-verzoek zodat deze
# gericht is op transparantie en niet op hergebruik. Hergebruik van 
# de broncode is toegestaan onder de EUPL licentie, met uitzondering 
# van broncode waarvoor een andere licentie is aangegeven.
# 
# Het archief waar dit bestand deel van uitmaakt is te vinden op:
#   https://github.com/MinBZK/woo-besluit-broncode-digid
# 
# Eventuele kwetsbaarheden kunnen worden gemeld bij het NCSC via:
#   https://www.ncsc.nl/contact/kwetsbaarheid-melden
# onder vermelding van "Logius, openbaar gemaakte broncode DigiD" 
# 
# Voor overige vragen over dit Woo-besluit kunt u mailen met open@logius.nl
# 
# This code has been disclosed in response to a request under the Dutch
# Open Government Act ("Wet open Overheid"). This implies that publication 
# is primarily driven by the need for transparence, not re-use.
# Re-use is permitted under the EUPL-license, with the exception 
# of source files that contain a different license.
# 
# The archive that this file originates from can be found at:
#   https://github.com/MinBZK/woo-besluit-broncode-digid
# 
# Security vulnerabilities may be responsibly disclosed via the Dutch NCSC:
#   https://www.ncsc.nl/contact/kwetsbaarheid-melden
# using the reference "Logius, publicly disclosed source code DigiD" 
# 
# Other questions regarding this Open Goverment Act decision may be
# directed via email to open@logius.nl

# frozen_string_literal: true

module RdwClient
  extend ActiveSupport::Concern

  included do
    helper_method :presentable_driving_licence if respond_to? :helper_method
  end

  def rdw_client
    Thread.current[:rdw_client] ||= DigidUtils::DhMu::RdwClient.new(
      endpoint: APP_CONFIG["urls"]["external"]["mu_endpoint"],
      open_timeout: ::Configuration.get_int("rdw_timeout"),
      read_timeout: ::Configuration.get_int("rdw_timeout"),
      ssl_cert_key_file: Rails.root.join(APP_CONFIG["rdw_ssl_cert_key_file"]&.path || APP_CONFIG["saml_ssl_cert_key_file"]&.path).to_s,
      ssl_cert_file: Rails.root.join(APP_CONFIG["rdw_ssl_cert_key_file"]&.path || APP_CONFIG["saml_ssl_cert_file"]&.path).to_s,
      ssl_ca_cert_file: APP_CONFIG["gba_ssl_ca_cert_file"]&.path.present? ? Rails.root.join(APP_CONFIG["gba_ssl_ca_cert_file"]&.path).to_s : nil,
      ssl_cert_key_password: Rails.application.secrets.private_key_password,
      ssl_verify_mode: :peer,
      log: Rails.env.test? ? false : true,
      namespaces: {"xmlns:a" => "http://www.w3.org/2005/08/addressing"}
    )
  end

  def fetch_driving_licences
    if current_account.bsn.present?
      rdw_client.get(bsn: current_account.bsn).tap do |driving_licences|
        session[:rdw_statuses] = driving_licences.map do |dl|
          { status: dl.human_status, document_no: dl.document_no, sequence_no: dl.sequence_no, canonical_type: "rijbewijs" }
        end
      end
    end
  end

  def presentable_driving_licence
    if tonen_rijbewijs_switch?
      session[:rdw_statuses]&.select{|dl| dl[:status] == "actief"}&.map{|card| OpenStruct.new(card.merge(canonical_type: "rijbewijs"))} || []
    else
      []
    end
  end

  def active_driving_licences_in_session?
    session[:rdw_statuses]&.select{|dl| dl[:status] == "actief"}.present?
  end
end
